File: Edu-JavaScript/src/com/jetbrains/edu/javascript/learning/JsCourseProjectGenerator.java
package com.jetbrains.edu.javascript.learning;

import com.intellij.javascript.nodejs.interpreter.NodeJsInterpreter;
import com.intellij.javascript.nodejs.interpreter.NodeJsInterpreterManager;
import com.intellij.javascript.nodejs.settings.NodeSettingsConfigurable;
import com.intellij.lang.javascript.modules.InstallNodeLocalDependenciesAction;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.application.ModalityState;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.module.ModuleManager;
import com.intellij.openapi.options.ShowSettingsUtil;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import com.jetbrains.edu.learning.OpenApiExtKt;
import com.jetbrains.edu.learning.courseFormat.Course;
import com.jetbrains.edu.learning.courseGeneration.GeneratorUtils;
import com.jetbrains.edu.learning.newproject.CourseProjectGenerator;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;

import static com.jetbrains.edu.learning.courseGeneration.GeneratorUtils.getInternalTemplateText;

public class JsCourseProjectGenerator extends CourseProjectGenerator<JsNewProjectSettings> {
  public static final Logger LOG = Logger.getInstance(JsCourseProjectGenerator.class);

  public JsCourseProjectGenerator(@NotNull JsCourseBuilder builder, @NotNull Course course) {
    super(builder, course);
  }

  @Override
  protected void afterProjectGenerated(@NotNull Project project, @NotNull JsNewProjectSettings projectSettings) {
    super.afterProjectGenerated(project, projectSettings);
    NodeJsInterpreter interpreter = projectSettings.getSelectedInterpreter();
    NodeJsInterpreterManager.getInstance(project).setInterpreterRef(interpreter.toRef());
    ModalityState modalityState = ModalityState.current();
    interpreter.provideCachedVersionOrFetch(version -> ApplicationManager.getApplication().invokeLater(() -> {
        if (version != null) {
          EduNodeJsCoreLibraryConfigurator.configureAndAssociateWithProject(project, interpreter, version);
        }
        else {
          LOG.warn("Couldn't retrieve Node interpreter version");
          VirtualFile requester = ModuleManager.getInstance(project).getModules()[0].getModuleFile();
          ShowSettingsUtil.getInstance().editConfigurable(project, new NodeSettingsConfigurable(project, requester, true));
        }
      }, modalityState, project.getDisposed()));
    }


  @Override
  protected void createAdditionalFiles(@NotNull Project project, @NotNull VirtualFile baseDir) throws IOException {
    String packageJson = "package.json";
    VirtualFile packageJsonFile = baseDir.findChild(packageJson);
    if (packageJsonFile == null && !myCourse.isStudy()) {
      final String templateText = getInternalTemplateText(packageJson);
      packageJsonFile = GeneratorUtils.createChildFile(baseDir, packageJson, templateText);
    }

    if (packageJsonFile != null && !OpenApiExtKt.isUnitTestMode()) {
      InstallNodeLocalDependenciesAction.runAndShowConsole(project, packageJsonFile);
    }
  }
}

File: Edu-JavaScript/src/com/jetbrains/edu/javascript/learning/JsCourseProjectGenerator.kt
package com.jetbrains.edu.javascript.learning

import com.intellij.javascript.nodejs.interpreter.NodeJsInterpreterManager
import com.intellij.javascript.nodejs.library.core.NodeCoreLibraryConfigurator
import com.intellij.javascript.nodejs.settings.NodeSettingsConfigurable
import com.intellij.lang.javascript.ui.NodeModuleNamesUtil
import com.intellij.openapi.application.ModalityState
import com.intellij.openapi.diagnostic.logger
import com.intellij.openapi.module.ModuleManager
import com.intellij.openapi.project.Project
import com.jetbrains.edu.learning.*
import com.jetbrains.edu.learning.courseFormat.Course
import com.jetbrains.edu.learning.courseGeneration.GeneratorUtils.createChildFile
import com.jetbrains.edu.learning.courseGeneration.GeneratorUtils.getInternalTemplateText
import com.jetbrains.edu.learning.newproject.CourseProjectGenerator
import java.io.IOException

class JsCourseProjectGenerator(builder: JsCourseBuilder, course: Course) : CourseProjectGenerator<JsNewProjectSettings>(builder, course) {
  override fun afterProjectGenerated(project: Project, projectSettings: JsNewProjectSettings, onConfigurationFinished: () -> Unit) {
    val interpreter = projectSettings.selectedInterpreter
    if (interpreter == null) {
      // It's ok not to have NodeJS interpreter in tests
      if (!isUnitTestMode) {
        LOG.warn("NodeJS interpreter is not selected")
      }
      return
    }
    NodeJsInterpreterManager.getInstance(project).setInterpreterRef(interpreter.toRef())

    val packageJsonFile = project.courseDir.findChild(NodeModuleNamesUtil.PACKAGE_JSON)
    // Don't install dependencies in headless mode (tests or course creation using `EduCourseCreatorAppStarter` on remote)
    // It doesn't make sense for tests since we don't check it.
    // On remote dependencies will be installed during warmup phase
    if (packageJsonFile != null && !isHeadlessEnvironment) {
      installNodeDependencies(project, packageJsonFile)
    }

    val modalityState = ModalityState.current()
    interpreter.provideCachedVersionOrFetch { version ->
      project.invokeLater(modalityState) {
        if (version != null) {
          val configurator = NodeCoreLibraryConfigurator.getInstance(project)
          configurator.configureAndAssociateWithProject(interpreter, version) {
            onConfigurationFinished()
          }
        }
        else {
          LOG.warn("Couldn't retrieve Node interpreter version")
          @Suppress("UnstableApiUsage")
          val requester = ModuleManager.getInstance(project).modules[0].moduleFile
          NodeSettingsConfigurable.showSettingsDialog(project, requester)
          onConfigurationFinished()
        }
      }
    }
    // Pass empty callback here because Core library configuration will be made asynchronously
    // Before this, we can't consider JS course project is fully configured
    super.afterProjectGenerated(project, projectSettings, onConfigurationFinished = {})
  }

  @Throws(IOException::class)
  override fun createAdditionalFiles(holder: CourseInfoHolder<Course>, isNewCourse: Boolean) {
    val packageJsonFile = holder.courseDir.findChild(NodeModuleNamesUtil.PACKAGE_JSON)
    if (packageJsonFile == null && !holder.course.isStudy) {
      val templateText = getInternalTemplateText(NodeModuleNamesUtil.PACKAGE_JSON)
      createChildFile(holder, holder.courseDir, NodeModuleNamesUtil.PACKAGE_JSON, templateText)
    }
  }

  companion object {
    private val LOG = logger<JsCourseProjectGenerator>()
  }
}